### PR TITLE
HypNative support

### PR DESCRIPTION
**Pull request by jmrossy**
- Add support for HypNative tokens
- Modify token list schema to reduce ambiguity
- Create global ChainId and DomainId type aliases for clarity 

Corresponds with https://github.com/hyperlane-xyz/hyperlane-deploy/pull/49

### CODE REVIEW

File: src/components/toast/TxSuccessToast.tsx [hunk 1]
```diff
@@ -2,7 +2,7 @@ import { toast } from 'react-toastify';
 
 import { getMultiProvider } from '../../features/multiProvider';
 
-export function toastTxSuccess(msg: string, txHash: string, chainId: number) {
+export function toastTxSuccess(msg: string, txHash: string, chainId: ChainId) {
```
Comment (asaj) [inline src/components/toast/TxSuccessToast.tsx:5]: Where is this coming from? I don't see it imported

File: src/features/contracts/hypErc20.ts [hunk 2]
```diff
@@ -1,17 +1,46 @@
...
+import {
+  HypERC20Collateral__factory,
+  HypERC20__factory,
+  HypNative__factory,
+  TokenType,
+} from '@hyperlane-xyz/hyperlane-token';
 
-export function getHypErc20CollateralContract(
+// Get the connected HypERC20Collateral, HypNative, or HypERC20 contract
+export function getHypWrapperContract(
```
Comment (asaj) [inline src/features/contracts/hypErc20.ts]: Should this be `getTokenRouterContract`? I believe that is what they all have in common

File: src/features/tokens/metadata.ts [hunk 3]
```diff
@@ -1,25 +1,48 @@
...
+        type: TokenType.collateral,
+        hypWrapperAddress: token.hypCollateralAddress,
+        address: token.address,
+      });
+    } else if (type == TokenType.native) {
+      tokenMetadata.push({
+        ...commonFields,
+        type: TokenType.native,
+        hypWrapperAddress: token.hypNativeAddress,
+        // Note, using 0x000... address to help identify native tokens
```
Comment (asaj) [inline src/features/tokens/metadata.ts:42]: Why is this necessary given `type`?

File: src/features/tokens/routes.ts [hunk 4]
```diff
@@ -1,35 +1,36 @@
...
-import { getHypErc20CollateralContract } from '../contracts/hypErc20';
+import { getHypWrapperContract } from '../contracts/hypErc20';
 import { getProvider } from '../multiProvider';
 
 import { getAllTokens } from './metadata';
 import { TokenMetadata, TokenMetadataWithHypTokens } from './types';
 
 export enum RouteType {
-  NativeToRemote = 'nativeToRemote',
+  BaseToRemote = 'baseToRemote',
```
Comment (asaj) [inline src/features/tokens/routes.ts]: nit: `base` is new terminology to me, usually we talk about local/remote or origin/destination.

File: src/features/tokens/routes.ts [hunk 5]
```diff
@@ -1,35 +1,36 @@
...
+  baseTokenAddress: Address;
+  hypWrapperAddress: Address;
   sourceTokenAddress: Address;
   destTokenAddress: Address;
   decimals: number;
 }
 
 // Source chain to destination chain to Route
-export type RoutesMap = Record<number, Record<number, Route[]>>;
+export type RoutesMap = Record<ChainId, Record<ChainId, Route[]>>;
```
Comment (asaj) [inline src/features/tokens/routes.ts:33]: nit: ditto above, s/source/origin

File: src/features/tokens/routes.ts [hunk 6]
```diff
@@ -57,35 +58,36 @@ export function useTokenRoutes() {
...
-    );
-  }
-  const symbolOnChain = await erc20.symbol();
-  if (symbol !== symbolOnChain) {
-    throw new Error(
-      `Token config symbol ${symbol} does not match contract decimals ${symbolOnChain}`,
-    );
+  const wrapperContract = getHypWrapperContract(type, hypWrapperAddress, provider);
+
+  if (type === TokenType.collateral) {
```
Comment (asaj) [inline src/features/tokens/routes.ts:66]: Is this not necessary for native as well?

File: src/features/tokens/routes.ts [hunk 7]
```diff
@@ -57,35 +58,36 @@ export function useTokenRoutes() {
...
   logger.info('Fetching connected domains');
-  const domains = await collateralContract.domains();
+  const domains = await wrapperContract.domains();
   logger.info(`Found ${domains.length} connected domains:`, domains);
 
   logger.info('Getting domain router address');
-  const hypTokenByteAddresses = await Promise.all(
-    domains.map((d) => collateralContract.routers(d)),
-  );
+  const hypTokenByteAddresses = await Promise.all(domains.map((d) => wrapperContract.routers(d)));
```
Comment (asaj) [inline src/features/tokens/routes.ts]: nit: hypTokenAddressesAsBytes32 or st

File: src/features/tokens/types.ts [hunk 8]
```diff
@@ -1,22 +1,83 @@
...
+
+const CollateralTokenSchema = commonTokenFields.extend({
+  type: z.literal(TokenType.collateral),
+  address: z.string(),
+  hypCollateralAddress: z.string(),
+});
 
-export const TokenListSchema = z.array(TokenSchema);
+interface CollateralTokenConfig extends BaseTokenConfig {
+  type: TokenType.collateral | 'collateral';
```
Comment (asaj) [inline src/features/tokens/types.ts:37]: I'm surprised we need to do this if the enum assigns `TokenType.collateral` to 'collateral'. Does it?

File: src/features/tokens/types.ts [hunk 9]
```diff
@@ -1,22 +1,83 @@
...
+  address: Address;
+  hypWrapperAddress: Address; // Shared name for hypCollateralAddr and hypNativeAddr
+}
+
+interface CollateralTokenMetadata extends BaseTokenMetadata {
+  type: TokenType.collateral;
+}
+
+type ZeroAddress = `${typeof ethers.constants.AddressZero}`;
+interface NativeTokenMetadata extends BaseTokenMetadata {
```
Comment (asaj) [inline src/features/tokens/types.ts:73]: This deserves a comment explaining what's going on here (as mentioned before, I don't understand why this is necessary)

File: src/features/tokens/useTokenBalance.tsx [hunk 10]
```diff
@@ -46,11 +48,18 @@ export function getCachedTokenBalance(
...
-async function fetchTokenBalance(chainId: number, tokenAddress: Address, accountAddress: Address) {
-  logger.debug(
-    `Fetching balance for account ${accountAddress} token ${tokenAddress} on chain ${chainId}`,
-  );
-  const erc20 = getErc20Contract(tokenAddress, getProvider(chainId));
-  const balance = await erc20.balanceOf(accountAddress);
-  return balance.toString();
+async function fetchTokenBalance(chainId: ChainId, tokenAddress: Address, accountAddress: Address) {
+  if (isNativeToken(tokenAddress)) {
+    logger.debug(`Fetching balance for account ${accountAddress} native token on chain ${chainId}`);
```
Comment (asaj) [inline src/features/tokens/useTokenBalance.tsx:53]: nit: if a MP is available it would be more readable to log the chain name

File: src/global.d.ts [hunk 11]
```diff
@@ -1 +1,3 @@
 declare type Address = string;
+declare type ChainId = number;
```
Comment (asaj) [inline src/global.d.ts:2]: Ah, there we go, I didn't know you could do this.

**Review by asaj — APPROVED** [review]
New to this repo so apologies if I'm off the mark on some things

File: src/features/tokens/routes.ts (hunk 4 above)
Comment (jmrossy) [reply src/features/tokens/routes.ts]: Yes but those terms are relative. I need a term that refers to the original source of the token. Totally open to renaming if you have a preferred alternative?

File: src/features/tokens/metadata.ts (hunk 3 above)
Comment (jmrossy) [reply src/features/tokens/metadata.ts:42]: It's not strictly necessary but I found that for other parts of the code (mostly UI code) it was helpful for *all* metadata objects to have an addresss

File: src/features/contracts/hypErc20.ts (hunk 2 above)
Comment (jmrossy) [reply src/features/contracts/hypErc20.ts]: That's what I meant by this yes. I'll rename

File: src/features/tokens/routes.ts (hunk 5 above)
Comment (jmrossy) [reply src/features/tokens/routes.ts:33]: Okay will rename tmrw

File: src/features/tokens/routes.ts (hunk 6 above)
Comment (jmrossy) [reply src/features/tokens/routes.ts:66]: What would we validate it against?

File: src/features/tokens/types.ts (hunk 8 above)
Comment (jmrossy) [reply src/features/tokens/types.ts:37]: Typescript won't allow a field of type `TokenType` to be the literal value 'collateral' even if the enum's values match up. And I'm determined to let devs use the literals in the configs to avoid them having to import `TokenType`. However I was not okay with loosening the type throughout the whole app. This separation of *Config and *Metadata types is the solution

File: src/features/tokens/types.ts (hunk 9 above)
Comment (jmrossy) [reply src/features/tokens/types.ts:73]: See above. I'll add a comment

File: src/features/tokens/useTokenBalance.tsx (hunk 10 above)
Comment (jmrossy) [reply src/features/tokens/useTokenBalance.tsx:53]: this is just a log statement, not user facing
